### PR TITLE
upgrade the interface of a2c and Alphazero algorithm

### PR DESCRIPTION
**Pull request by wangzelong0663**
1.modify the actor with '@parl.remote_class(wait=False).'
2.modify the coach.py with the new interface.

### CODE REVIEW

File: benchmark/torch/AlphaZero/Coach.py [hunk 1]
```diff
@@ -120,12 +76,17 @@ def learn(self):
             logger.info('Step1: self-play in parallel...')
             iterationTrainExamples = []
             # update weights of remote actors to the latest weights, and ask them to run self-play task
-            for signal_queue in self.remote_actors_signal_queues:
-                signal_queue.put({"task": "self-play"})
-            # wait for all remote actors (a total of self.args.actors_num) to return the self-play results
-            for _ in range(self.args.actors_num):
-                result = self.remote_actors_return_queue.get()
-                iterationTrainExamples.extend(result["self-play"])
+            # and get the total self play data example of all the actors for training.
```
Comment (zenghsh3) [inline benchmark/torch/AlphaZero/Coach.py:79]: Maybe we can rewrite this sentence, it's not very smooth.

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
LGTM.

File: benchmark/torch/AlphaZero/Coach.py [hunk 2]
```diff
@@ -120,12 +76,17 @@ def learn(self):
...
+            episode_num_each_actor = self.args.numEps // self.args.actors_num
+
+            future_object_ids  = [remote_actor.self_play(
+                self.current_agent.get_weights(), episode_num_each_actor) \
```
Comment (zenghsh3) [inline benchmark/torch/AlphaZero/Coach.py:83]: Optimize the logic:
```
weights = self.current_agent.get_weights()
future_object_ids  = [remote_actor.self_play(weights, episode_num_each_actor)
```

File: benchmark/torch/AlphaZero/Coach.py [hunk 3]
```diff
@@ -155,23 +116,24 @@ def learn(self):
...
-                    "task":
-                    "evaluate_test_dataset",
-                    "test_dataset":
-                    data
-                })
+                split_datas.append(data)
                 cnt += len(data)
+
+            future_object_ids  = [remote_actor.evaluate_test_dataset(
+                self.current_agent.get_weights(), data) \
```
Comment (zenghsh3) [inline benchmark/torch/AlphaZero/Coach.py:128]: Optimize the logic，same as above.

File: benchmark/torch/AlphaZero/Coach.py [hunk 4]
```diff
@@ -185,12 +147,18 @@ def learn(self):
             logger.info(
                 'Step4: pitting against previous generation in parallel...')
             # transfer weights of previous generation and current generation to the remote actors, and ask them to pit.
-            for signal_queue in self.remote_actors_signal_queues:
-                signal_queue.put({"task": "pitting"})
+            games_num_each_actor = self.args.arenaCompare // self.args.actors_num
+            future_object_ids  = [remote_actor.pitting(
+                self.previous_agent.get_weights(),
```
Comment (zenghsh3) [inline benchmark/torch/AlphaZero/Coach.py:152]: Optimize the logic，same as above.

File: benchmark/torch/a2c/train.py [hunk 5]
```diff
@@ -78,59 +76,41 @@ def __init__(self, config, cuda):
...
+        1.setting latest_params to each actor model
+        2.getting the sample data from all the actors synchronizely
+        3.traing the model with the sample data and the params is upgraded, and goto step 1
+        """
 
-        for params_queue in self.params_queues:
-            params_queue.put(latest_params)
+        latest_params = self.agent.get_weights()
+        # setting the actor to the latest_params
+        [
```
Comment (zenghsh3) [inline benchmark/torch/a2c/train.py:100]: We don't need to use the list iterator here.

File: benchmark/torch/a2c/train.py [hunk 6]
```diff
@@ -232,8 +216,11 @@ def should_stop(self):
...
     assert config['log_metrics_interval_s'] > 0
 
+    cnt = 1
     while not learner.should_stop():
         start = time.time()
         while time.time() - start < config['log_metrics_interval_s']:
             learner.step()
-        learner.log_metrics()
+            cnt += 1
+            if cnt % learner.config['get_remote_metrics_interval'] == 0:
```
Comment (zenghsh3) [inline benchmark/torch/a2c/train.py:225]: Maybe we can remove this condition.

**Pull request closed without merging**